File: projector-ts/src/projector.ts
import type { Config } from "./config"
import * as fs from "node:fs"
import * as path from "node:path"
export type Data = {
    projector: {
        //pwd
        [key: string]: {
            // key -> value
            [key: string]: string
        }
    }
}

const defaultData = {
    projector: {}
}

export default class Projector {
    constructor(private config: Config, private data: Data) { }

    getValueAll(): { [key: string]: string } {
        let curr = this.config.pwd;
        let prev = "";

        const paths = [];
        do {
            prev = curr;
            paths.push(curr)
            curr = path.dirname(curr);
        } while (curr != prev)

        return paths.reverse().reduce((acc, path) => {
            const value = this.data.projector[path];
            if (value) {
                Object.assign(acc, value);
            }

            return acc;
        }, {})
    }


    getValue(key: string): string | undefined {
        let curr = this.config.pwd;
        let prev = "";
        let out = undefined
        do {
            const value = this.data.projector[curr]?.[key];
            if (value) {
                out = value;
                break
            }

            prev = curr;
            curr = path.dirname(curr);
        } while (curr != prev)

        return out;
    }

    setValue(key: string, value: string) {
        let dir = this.data.projector[this.config.pwd];
        if (!dir) {
            dir = this.data.projector[this.config.pwd] = {}
        }

        dir[key] = value;
    }
    
    removeValue(key: string) {
        const dir = this.data.projector[this.config.pwd];
        if (dir) {
            delete dir[key];
        }
    }

    save() {
        const configPath = path.dirname(this.config.config);
        if (!fs.existsSync(configPath)) {
            fs.mkdirSync(configPath, {recursive: true})
        }

        fs.writeFileSync(this.config.config, JSON.stringify(this.data))
    }

    static FromConfig(config: Config): Projector {
        if (fs.existsSync(config.config)) {
            let data: Data;
            try {
                data = JSON.parse(fs.readFileSync(config.config).toString())
            } catch (error) {
                data = defaultData
            }
            return new Projector(config, data)
        }

        return new Projector(config, defaultData)
    }
}

File: projector-rust/src/projector.rs
use std::{collections::HashMap, path::PathBuf};

use serde::{Deserialize, Serialize};

use crate::config::Config;

#[derive(Debug, Serialize, Deserialize)]
pub struct Data {
    pub projector: HashMap<PathBuf, HashMap<String, String>>,
}

pub struct Projector {
    config: PathBuf,
    pwd: PathBuf,
    data: Data,
}

fn default_data() -> Data {
    return Data {
        projector: HashMap::new(),
    };
}

impl Projector {
    pub fn get_value_all(&self) -> HashMap<&String, &String> {
        let mut curr = Some(self.config.pwd.as_path());
        let mut paths = vec![];
        while let Some(p) = curr {
            paths.push(p);
            curr = p.parent();
        }

        let mut out = HashMap::new();

        for path in paths.into_iter().rev() {
            if let Some(map) = self.data.projector.get(&PathBuf::from(path)) {
                out.extend(map.iter());
            }
        }

        return out;
    }

    pub fn set_value(&mut self, key: String, value: String) {
        self.data
            .projector
            .entry(self.pwd.clone())
            .or_default()
            .insert(key, value);
    }

    pub fn remove_value(&mut self, key: &str) {
        if let Some(dir) = self.data.projector.get_mut(&self.pwd) {
            dir.remove(key);
        }
    }

    pub fn get_value(&self, key: String) -> Option<&String> {
        let mut curr = Some(self.pwd.as_path());
        let mut out = None;
        while let Some(p) = curr {
            if let Some(dir) = self.data.projector.get(&PathBuf::from(p)) {
                if let Some(value) = dir.get(&key) {
                    out = Some(value);
                    break;
                }
            }
            curr = p.parent();
        }
        return out;
    }

    pub fn from_config(config: PathBuf, pwd: PathBuf) -> Self {
        if std::fs::metadata(&config).is_ok() {
            let contents = std::fs::read_to_string(&config.config);

            let contents = contents.unwrap_or(String::from("{\"projector\":{}}"));
            let data = serde_json::from_str(&contents);
            let data = data.unwrap_or(default_data());

            return Projector { config, pwd, data };
        }

        return Projector {
            config,
            pwd,
            data: default_data(),
        };
    }

    pub fn save(&self) -> Result<()> {
        if let Some(p) = self.config.parent() {
            if !std::fs::metadata(&p).is_ok() {
                std::fs::create_dir_all(p)?;
            }
        };

        let contents = serde_json::to_string(&self.data)?;
        std::fs::write(&self.config, contents)?;

        return Ok(());
    }
}

#[cfg(test)]
mod test {

    use std::{collections::HashMap, path::PathBuf};

    use anyhow::Ok;
    use collection_macros::hashmap;

    use crate::{
        config::{Config, Operation},
        projector::{Data, Projector},
    };

    fn get_data() -> HashMap<PathBuf, HashMap<String, String>> {
        return hashmap! {
                PathBuf::from("/") => hashmap! {
                    "foo".into() => "bar1".into(),
                    "bar".into() => "bazz".into(),
                },
                PathBuf::from("/foo") => hashmap! {
                    "foo".into() => "bar2".into()
                },
                PathBuf::from("/foo/bar") => hashmap! {
                    "foo".into() => "bar3".into()
                },
                PathBuf::from("/foo/bar/baz") => hashmap! {
                    "foo".into() => "bar3".into()
                },
        };
    }

    fn get_projector(pwd: PathBuf) -> Projector {
        return Projector {
            config: PathBuf::from(""),
            pwd,
            data: Data {
                projector: get_data(),
            },
        };
    }

    #[test]
    fn get_value() {
        let proj = get_projector(PathBuf::from("/foo/bar"));

        assert_eq!(proj.get_value("foo".into()), Some(&String::from("bar3")));
    }
}
